src/parser/SpatialQuery.cpp

```diff
@@ -60,7 +64,25 @@ void SpatialQuery::addParameter(const SparqlTriple& triple) {
           "<baseline>, <s2> or <boundingBox>.");
     }
   } else if (predString == "payload") {
-    payloadVariables_.push_back(getVariable("payload", object));
+    if (object.isVariable()) {
+      // Single selected variable
+
+      // If we have already selected all payload variables, we can ignore
+      // another explicitly selected variable.
+      if (std::holds_alternative<std::vector<Variable>>(payloadVariables_)) {
+        std::get<std::vector<Variable>>(payloadVariables_)
+            .push_back(getVariable("payload", object));
+      }
+    } else if (object.isIri() &&
+               extractParameterName(object, SPATIAL_SEARCH_IRI) == "all") {
+      // All variables selected
+      payloadVariables_ = PayloadAllVariables{};
+    } else {
+      throw SpatialSearchException(
+          "The argument to the <payload> parameter must be either a variable "
+          "to be selected or <all>.");
+    }
+
```

Comment:
The `PayloadVariables` now have quite some semantics.
I think we get the cleanest code, if we extract them to a separate class in a separate file, s.t. you can write something like 

```
PayloadVariables payload = PayloadVariables::emptyPayload();
payload.addVar(variable); // Do nothing if allready specified to `all`.
payload.setToAll();
payload.empty();
payload.isAll();
payload.getVariables(); // Throw if `isAll`.
```

That way, you can remove some dependencies (e.g. the SpatialJoin.h header), and the code gets much more readable
(self documenting).